web/src/unread.js

```diff
@@ -691,8 +693,12 @@ export function update_message_for_mention(message, content_edited = false) {
     if (is_unmuted_mention || message.mentioned_me_directly) {
         unread_mentions_counter.add(message.id);
         add_message_to_unread_mention_topics(message.id);
+        if (!message.stream_id) {
+            mentions_in_direct_message.add(message.id);
+        }
     } else {
         unread_mentions_counter.delete(message.id);
+        mentions_in_direct_message.delete(message.id);
         remove_message_from_unread_mention_topics(message.id);
     }
```

Comment:
How does this logic handle an `@all` or `@group` that mentions the current user in a direct message? I wonder if the `is_unmuted_mention` calculation needs to be adjusted to be something like `message.mentioned && message.type !== stream || !user_topics.is_topic_muted(message.stream_id, message.topic);`

Reply:
This logic handles `@group` well in direct messages. 

But for `@all` and `@everyone` they are not being counted in the mention unread count at all when in direct messages, although they are shown in the mentions view and also are highlighted as mentions, is it an Issue that needs to be solved? 

![image](https://github.com/zulip/zulip/assets/64723994/5ec341cf-4a80-4e0f-9913-591b8eea37e7)

Reply:
Is the reasoning that `unread_mentions_counter` doesn't get those wildcard mentions, so `direct_message_with_mention_count` doesn't need them?

Reply:
@timabbott Here is an explanation, it is a little long😅, hope I explained it clearly:

<details>
  <summary>Code</summary>
  ```
const is_unmuted_mention =
        message.type === "stream" &&
        message.mentioned &&
        !user_topics.is_topic_muted(message.stream_id, message.topic);

    if (is_unmuted_mention || message.mentioned_me_directly) {
        unread_mentions_counter.add(message.id);
        add_message_to_unread_mention_topics(message.id);
        if (!message.stream_id) {
            direct_message_with_mention_count.add(message.id);
        }
    }
```
</details>

In the `update_message_for_mention` function, `unread_mentions_counter` is increased if `is_unmuted_mention` is true or `message.mentioned_me_directly` is true and `direct_message_with_mention_count` is increased too but only if the message is a direct message.

so for `@user` and `@group` mentions in direct messages, `direct_message_with_mention_count` is increased as `message.mentioned_me_directly` is true for them. 

But for `@all` and `@everyone` mentions in direct messages, `direct_message_with_mention_count` is not increased as 
`message.mentioned_me_directly` is false and `is_unmuted_mention` is also false as `message.type != "stream"` in `is_unmuted_mention` calculation.

And you are correct that the reasoning is that `unread_mentions_counter` also doesn't get wildcard mentions, in fact, wildcard mentions in direct mentions are not counted as mentions at all, which is a bug I have also reported on [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Wildcard.20mentions.20in.20direct.20messages/near/1641123).

But they are not considered as mentions only until the page reloads due to this bug, but after reloading they are counted as mentions, as after reloading counting is done in the `initialize()` function of `unread.js`, which handles all the mentions correctly and also increases `direct_message_with_mention_count` correctly. That's why I commented that it handles all the cases correctly and the screenshot is also after page reloading.